neqo-transport/src/connection/mod.rs

```diff
@@ -1287,17 +1288,19 @@ impl Connection {
 
     /// In case a datagram arrives that we can only partially process, save any
     /// part that we don't have keys for.
-    fn save_datagram(&mut self, cspace: CryptoSpace, d: &Datagram, remaining: usize, now: Instant) {
-        let d = if remaining < d.len() {
-            Datagram::new(
-                d.source(),
-                d.destination(),
-                d.tos(),
-                &d[d.len() - remaining..],
-            )
-        } else {
-            d.clone()
-        };
+    fn save_datagram(
+        &mut self,
+        cspace: CryptoSpace,
+        d: &Datagram<impl AsRef<[u8]>>,
+        remaining: usize,
+        now: Instant,
+    ) {
+        let d = Datagram::new(
+            d.source(),
+            d.destination(),
+            d.tos(),
+            d[d.len() - remaining..].to_vec(),
+        );
```

Comment:
Now always calls `to_vec`, even if the full datagram payload should be saved. `Datagram<impl AsRef<[u8]>>` can not differentiate in a `&[u8]` payload and a `Vec<u8>`, i.e. does not know when to allocate (`&[u8]`) and when to return itself (`Vec<u8>`).

Maybe the compiler optimizes this away.

I have to give this more thought.

Reply:
If you remove the reference in the signature, that would help, I think.  Otherwise, the compiler won't be able to observe that the owned datagram is really owned.  The problem is that there is a slice involved, which might not easily map to just taking the owned datagram and transferring it (rather than allocating).  That is a common case, but maybe not something that the compiler can detect reliably.

I'm not concerned too much about the extra allocation either way. It only happens during the handshake and not that often. From memory, it's not exclusively the case that loss or reordering lead to the need to save datagrams, but those are the most obvious reasons for needing to save. Both are rare enough that we shouldn't worry about an allocation or two, especially when those allocations are pretty cheap.

In short, don't sweat it.  If it bothers you, open an issue so that we can track the problem and come back to it.